File: 백준/Gold/1275. 커피숍2/커피숍2.java
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.StreamTokenizer;

class Main {
    static StreamTokenizer sttk = new StreamTokenizer(new InputStreamReader(System.in));
    static long[] tree;
    static int n, s;

    static int readInt() throws IOException {
        sttk.nextToken();
        return (int) sttk.nval;
    }

    public static void main(String[] args) throws IOException {
        n = readInt();
        int q = readInt();

        init();

        StringBuilder sb = new StringBuilder();

        while (q-- > 0) {
            sb.append(query(readInt(), readInt())).append('\n');
            update(readInt(), readInt());
        }

        System.out.println(sb);
    }

    static void init() throws IOException {
        s = 1;

        while (s < n) s <<= 1;

        tree = new long[s << 1];

        for (int i = s; i < s + n; i++) {
            tree[i] = readInt();
        }

        for (int i = s - 1; i > 0; i--) {
            tree[i] = tree[i << 1] + tree[(i << 1) + 1];
        }
    }

    static void update(int idx, int v) {
        int node = s + idx - 1;
        long diff = v - tree[node];

        while (node > 0) {
            tree[node] += diff;
            node >>>= 1;
        }
    }

    static long query(int qs, int qe) {
        if (qs > qe) {
            return query(qe, qs);
        }

        qs += s - 1;
        qe += s - 1;

        long ret = 0;

        while (qs <= qe) {
            if ((qs & 1) == 1) ret += tree[qs++];

            if ((qe & 1) == 0) ret += tree[qe--];

            qs >>>= 1;
            qe >>>= 1;
        }

        return ret;
    }
}

File: 백준/Gold/1275. 커피숍2/커피숍2.kt
import java.io.StreamTokenizer

val sttk = StreamTokenizer(System.`in`.bufferedReader())

fun readInt() = with(sttk) {
    nextToken()
    nval.toInt()
}

fun main() {
    val n = readInt()
    val q = readInt()
    val segTree = SegmentTree(n)

    repeat(n) {
        segTree.insert(1, n, 1, it + 1, readInt().toLong())
    }

    val sb = StringBuilder()

    repeat(q) {
        val sum = segTree.query(readInt(), readInt())

        sb.append(sum).append('\n')

        segTree.update(1, n, 1, readInt(), readInt().toLong())
    }

    print(sb)
}

class SegmentTree(val n: Int) {
    val tree = LongArray(n shl 2)

    tailrec fun insert(s: Int, e: Int, node: Int, idx: Int, v: Long) {
        tree[node] += v

        if (s == e) return

        val mid = (s + e) ushr 1
        val child = node shl 1

        if (idx in s..mid) {
            insert(s, mid, child, idx, v)
        } else {
            insert(mid + 1, e, child + 1, idx, v)
        }
    }

    fun update(s: Int, e: Int, node: Int, idx: Int, v: Long) {
        if (idx !in s..e) return

        if (s == e) {
            tree[node] = v
            return
        }

        val mid = (s + e) ushr 1
        val child = node shl 1

        update(s, mid, child, idx, v)
        update(mid + 1, e, child + 1, idx, v)

        tree[node] = tree[child] + tree[child + 1]
    }

    fun query(s: Int, e: Int): Long {
        return if (s < e)
            query(1, n, 1, s, e)
        else
            query(1, n, 1, e, s)
    }

    private fun query(s: Int, e: Int, node: Int, qs: Int, qe: Int): Long {
        if (s > qe || e < qs) return 0L

        if (qs <= s && e <= qe) return tree[node]

        val mid = (s + e) ushr 1
        val child = node shl 1

        return query(s, mid, child, qs, qe) + query(mid + 1, e, child + 1, qs, qe)
    }
}